larpandoradlcontent/LArDeepLearning/DeepLearningTrackShowerIdAlgorithm.cc

```diff
@@ -0,0 +1,352 @@
+/**
+ *  @file   larpandoradlcontent/LArDeepLearning/DeepLearningTrackShowerIdAlgorithm.cc
+ *
+ *  @brief  Implementation of the deep learning track shower id algorithm.
+ *
+ *  $Log: $
+ */
+
+#include "Pandora/AlgorithmHeaders.h"
+
+#include <torch/script.h>
+#include <torch/torch.h>
+
+#include "larpandoradlcontent/LArDeepLearning/DeepLearningTrackShowerIdAlgorithm.h"
+#include "larpandoradlcontent/LArHelpers/LArDLHelper.h"
+#include "larpandoracontent/LArHelpers/LArFileHelper.h"
+#include "larpandoracontent/LArHelpers/LArMCParticleHelper.h"
+#include "larpandoracontent/LArHelpers/LArMvaHelper.h"
+#include "larpandoracontent/LArHelpers/LArMonitoringHelper.h"
+#include "larpandoracontent/LArHelpers/LArPfoHelper.h"
+
+#include <chrono>
+
+using namespace pandora;
+
+namespace lar_content
+{
+
+DeepLearningTrackShowerIdAlgorithm::DeepLearningTrackShowerIdAlgorithm() :
+    m_visualize(false),
+    m_useTrainingMode(false),
+    m_profile(false),
+    m_trainingOutputFile("")
+{
+}
+
+DeepLearningTrackShowerIdAlgorithm::~DeepLearningTrackShowerIdAlgorithm()
+{
+    if (m_profile)
+    {
+        try
+        {
+            PANDORA_MONITORING_API(SaveTree(this->GetPandora(), "cpu_tree", "cpu.root", "UPDATE"));
+        }
+        catch(const StatusCodeException&)
+        {
+            std::cout << "DeepLearningTrackShowerIdAlgorithm: Unable to write tree to file" << std::endl;
+        }
+    }
+}
+
+StatusCode DeepLearningTrackShowerIdAlgorithm::Run()
+{
+    if (m_useTrainingMode)
+        return Train();
+    else
+        return Infer();
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+StatusCode DeepLearningTrackShowerIdAlgorithm::Train()
+{
+    const int SHOWER{1}, TRACK{2};
+    for (const std::string listName : m_caloHitListNames)
+    {
+        const CaloHitList *pCaloHitList(nullptr);
+        PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::GetList(*this, listName, pCaloHitList));
+        const MCParticleList *pMCParticleList(nullptr);
+        PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::GetCurrentList(*this, pMCParticleList));
+
+        const bool isU(pCaloHitList->front()->GetHitType() == TPC_VIEW_U ? true : false);
```

Comment:
Slightly cleaner to just cache and check the HitType and work with it, rather than these Booleans? (Little difference, really, of course.)